src/keywrap/pkcs11/mod.rs

```diff
@@ -0,0 +1,150 @@
+        for p in ps {
+            x.push(p.to_vec());
+        }
+        for y in &ec.param["pkcs11-yamsl"] {
+            x.push(y.to_vec());
+        }
+        let pkcs11_recipients: Vec<Pkcs11KeyFileObject>
+          = add_pub_keys(&(ec.decrypt_config.as_ref().unwrap()), &x).unwrap();
+
+        if pkcs11_recipients.len() == 0 {
+            return Ok(Vec::new())
+        }
+
+        let json_str = encrypt_multiple(&pkcs11_recipients, opts_data).unwrap();
+
+        Ok(json_str)
+    }
+
+    fn unwrap_key(&self,
+                  dc: &DecryptConfig,
+                  annotation: &Vec<u8>)
+                  -> Result<Vec<u8>, std::io::Error> {
+        // TODO
+        Ok(b"".to_vec())
+    }
+
+
+    fn get_annotation_id(&self) -> &str {
+        "org.opencontainers.image.enc.keys.pkcs11"
+    }
+
+    fn no_possible_keys(&self,
+                        dcparameters: &HashMap<String, Vec<Vec<u8>>>)
+                        -> bool {
+        self.get_private_keys(dcparameters).len() == 0
+    }
+
+    fn get_private_keys<'a>(&self,
+                            dcparameters: &'a HashMap<String, Vec<Vec<u8>>>)
+                            -> &'a Vec<Vec<u8>> {
+        &dcparameters["pkcs11-yamls"]
+    }
+
+    fn get_key_ids_from_packet(&self,
+                               packet: String)
+                               -> Result<Vec<u64>, std::io::Error> {
+        // FIXME return nil, nil
+        Ok(Vec::new())
+    }
+
+
+    fn get_recipients(&self,
+                      packet: String)
+                      -> Result<Vec<String>, std::io::Error> {
+        let mut v = Vec::new();
+        v.push("[pkcs11]".to_string());
+        Ok(v)
+    }
+}
+
+
+fn p11_conf_from_params(dcparameters: &HashMap<String, Vec<Vec<u8>>>)
+                        -> Result<Pkcs11Config, std::io::Error> {
+    // FIXME: c is just a placeholder for now
+    let c = Pkcs11Config{
+        module_directories: Vec::default(),
+        allowed_module_paths: Vec::default(),
+    };
+    if dcparameters.contains_key("pkcs11-config") {
+        // TODO
+        //return pkcs11.ParsePkcs11ConfigFile(dcparameters["pkcs11-config"][0])
+        return Ok(c);
+    }
+    // FIXME was "nil, nil" in golang. Should probably be Option here
+    Ok(c)
+}
+
+fn add_pub_keys(dc: &DecryptConfig,
+                pub_keys: &Vec<Vec<u8>>)
+                -> Result<Vec<Pkcs11KeyFileObject>, std::io::Error> {
+    let mut pkcs11_keys = Vec::new();
+    if pub_keys.len() == 0 {
+        return Ok(pkcs11_keys);
+    }
+
+    let p11_conf = p11_conf_from_params(&dc.param).unwrap();
+
+    for k in pub_keys {
+            let key: Pkcs11KeyFileObject = parse_pkcs11_key_file(k).unwrap();
+            // FIXME: Do we need more fields for the key here?
+            //key.uri.SetModuleDirectories(p11conf.ModuleDirectories);
+            //key.uri.SetAllowedModulePaths(p11conf.AllowedModulePaths);
+            pkcs11_keys.push(key);
+    }
+
+    Ok(pkcs11_keys)
+}
+
```

Comment:
ik this function is not done yet, but it seems like maybe a good example of a very c-like approach. rather than individually adding things to a mutable vector, you might be able to return a match statement with `pub_keys.is_empty()` and either an empty vector or some map.

Reply:
I updated this based on the suggestion. I also handled the similar problem in unwrap_keys. The version I came up with is harder for me to read. That could be because I'm not used to Rust or because my implementation is still bad.